Bluetooth: Fix RX thread deadlock

joerchan opened the pull request:
Fix host RX thread being deadlocked. The deadlock occurs because the
RX thread is stuck waiting in conn_tx_alloc with K_FOREVER but if the
connection is disconnected only the RX thread can unblock it in the
handling of the disconnect event.

Fixes this deadlock by splitting the processing of the
disconnected event into two parts.
The part needed to unblock the RX is to release resources held by
unack'ed TX packets and mark the connection state as not connected
anymore.
The RX thread waiting for free_tx fifo and the TX thread waiting for
the bt_dev.le.pkts semaphore will both check the connected state after
having acquired them and will abort if disconnected.
The rest of the processing will be handled at normal RX thread
priority like normal.

The HCI drivers need to be updated to deliver the disconnected event
from both priority and normal RX context.

Also fix an issue with the advertiser not resuming connectable
advertiser if the last bt_conn_unref happens from something other
than the disconnected callback.
In this case this last bt_conn_unref was from gatt.c notify_cb called
from application main thread.


Fixes: #23364
Fixes: #26467

joerchan committed:
Bluetooth: Fix host RX thread deadlock

Fix host RX thread being deadlocked. The deadlock occurs because the
RX thread is stuck waiting in conn_tx_alloc with K_FOREVER but if the
connection is disconnected only the RX thread can unblock it in the
handling of the disconnect event.

This commit fixes this deadlock by splitting the processing of the
disconnected event into two parts.
The part needed to unblock the RX is to release resources held by
unack'ed TX packets and mark the connection state as not connected
anymore.
The RX thread waiting for free_tx fifo and the TX thread waiting for
the bt_dev.le.pkts semaphore will both check the connected state after
having acquired them and will abort if disconnected.
The rest of the processing will be handled at normal RX thread
priority like normal.

Move the bt_recv_prio handling to the Bluetooth host when the host
has defined its own RX thread (CONFIG_BT_RECV_IS_RX_THREAD=n).
If the HCI driver has the RX thread (CONFIG_BT_RECV_IS_RX_THREAD=y),
then the responsibility to call bt_recv and bt_recv_prio correctly
falls to the HCI driver.
The helper function bt_hci_evt_is_prio() is replaced with
bt_hci_evt_get_flags() so that the HCI driver can do this correctly.
This decision to replace was made so that existing HCI drivers
maintained out-of-tree will fail at compile time with the new system.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>
Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

Bluetooth: host: Move bt_recv_prio to host when RX thread is defined

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Fix resume advertiser race condition

This patch fixes an issue with the advertiser not resuming connectable
advertiser if the last bt_conn_unref happens from something other
than the disconnected callback.
In this case this last bt_conn_unref was from gatt.c notify_cb called
from application main thread.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

carlescufi commented:
@erwango @cvinayak @thoh-ot @alexandru-porosanu-nxp could you please test this with your respective Bluetooth platforms please? At the very least connection and disconnection multiple times, ideally with ongoing traffic.

Vudentz commented on subsys/bluetooth/host/conn_internal.h:
```diff
@@ -9,6 +9,7 @@
  */
 typedef enum __packed {
 	BT_CONN_DISCONNECTED,
+	BT_CONN_DISCONNECTED_PRIO,
```
I wonder if we coultn't just reuse BT_CONN_DISCONNECT for this, at least that would be consistent with the state when we initiate the disconnect, or we need to distinct of remote vs locally initiated?

joerchan replied:
For locally initiated it would be in the "DISCONNECT" already then, when receiving the HCI_DISCONNECT_COMPLETE event, so it would end up with a same state transition. `bt_conn_set_state` would return immediately in that case.

Vudentz commented on subsys/bluetooth/host/conn.c:
```diff
@@ -1774,8 +1774,11 @@ void bt_conn_set_state(struct bt_conn *conn, bt_conn_state_t state)
 		case BT_CONN_DISCONNECTED:
-			/* Cannot happen, no transition. */
+			/* Cannot happen. */
+			BT_WARN("no valid (%u) old state", state);
```
Id replace no valid/invalid

carlescufi replied:
Agreed, please replace `no valid` with `invalid`

alexandru-porosanu-nxp commented:
> @erwango @cvinayak @thoh-ot @alexandru-porosanu-nxp could you please test this with your respective Bluetooth platforms please? At the very least connection and disconnection multiple times, ideally with ongoing traffic.

Yep, I can do that; I'm thinking of using peripheral_hr (VEGA) & central_hr (NRF) and hack one of them to disconnect and reconnect at about 5-10s interval. If you have a better scenario in mind, let me know and I'll try to figure out how to do it...

cvinayak commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -272,6 +272,8 @@ static inline void read_payload(void)
 	}
 
 	prio = (rx.type == H4_EVT && bt_hci_evt_is_prio(rx.evt.evt));
+	normal = (rx.type == H4_ACL ||
```
Alternative suggestion, use a recv type and demux to both bt_recv_prio and bt_recv as suitable.

```
enum bt_recv_type {
    BT_RECV_TYPE_NORMAL = 0x00,
    BT_RECV_TYPE_PRIO = 0x02,
    BT_RECV_TYPE_BOTH = 0x03
}

u8_t bt_hci_evt_recv_type(u8_t evt)
{
    switch (evt) {
    case BT_HCI_EVT_CMD_COMPLETE:
    case BT_HCI_EVT_CMD_STATUS:
        /* fallthrough */
#if defined(CONFIG_BT_CONN)
    case BT_HCI_EVT_NUM_COMPLETED_PACKETS:
    case BT_HCI_EVT_DATA_BUF_OVERFLOW:
#endif
        return BT_RECV_TYPE_PRIO;

#if !defined(CONFIG_BT_HCI_RAW)
    case BT_HCI_EVT_DISCONN_COMPLETE:
        return BT_RECV_TYPE_BOTH;
#endif
    }
    return BT_RECV_TYPE_NORMAL;
}
```

in the drivers:
```
if (rx.type == H4_EVT) {
    recv_type = bt_hci_evt_recv_type(rx.evt.evt);
} else {
    recv_type = BT_RECV_TYPE_NORMAL;
}

switch (recv_type) {
case BT_RECV_TYPE_PRIO:
    bt_recv_prio(buf);
    break;
case BT_RECV_TYPE_BOTH:
    bt_recv_prio(buf);
    /* fall through */
case BT_RECV_TYPE_NORMAL:
default:
    bt_recv(buf);
    break;
}
```

carlescufi replied:
At some point before the PR was posted I suggested the same, and I think I still prefer that now, so I agree with you @cvinayak. 
I'd like to get @jhedberg's opinion as well.

carlescufi replied:
That said I think instead of an enum we could use a mask to simplify checks?

```
#define BT_EVT_TYPE_PRIO BIT(0)
#define BT_EVT_TYPE_NORMAL BIT(1)

if (type & BT_EVT_TYPE_PRIO) bt_recv_prio();
if (type & BT_EVT_TYPE_NORMAL) bt_recv_normal();

cvinayak commented on subsys/bluetooth/controller/hci/hci_driver.c:
```diff
@@ -105,6 +123,8 @@ static void prio_recv_thread(void *p1, void *p2, void *p3)
 		}
 
 		if (node_rx) {
+			bool evt_is_normal = true;
```
Where is this assigned to false?

cvinayak commented on subsys/bluetooth/controller/hci/hci_driver.c:
```diff
@@ -113,15 +133,24 @@ static void prio_recv_thread(void *p1, void *p2, void *p3)
 			node_rx->hdr.user_meta = hci_get_class(node_rx);
 
-			buf = process_prio_evt(node_rx);
+			buf = process_prio_evt(node_rx, &evt_is_normal);
```
returning a bt_recv_type and handling the call to bt_recv_prio or bt_recv in a switch case is cleaner.

joerchan replied:
@carlescufi @jhedberg opinions on removing bt_recv_is_prio and and bt_recv_is_normal with a bt_recv_type as suggested?
I prefer my way, but not a strong opinion.
Related comment: https://github.com/zephyrproject-rtos/zephyr/pull/25954#discussion_r435038613

joerchan replied:
I prefer if (is_normal) over if (type == normal || type == both)

carlescufi replied:
I prefer a bitmask, see my previous comment.
@jhedberg feedback welcome here as well.

cvinayak reviewed (dismissed):
I believe this PR fixes the stall issue, as I can see `recv_thread` loop twice under the scenarios where applications main thread and host's thread where waiting for tx_free counts. Previously, `recv_thread` would only run with k_fifo_get in conn_tx_alloc was updated to have a timeout.

erwango commented:
Testing on nucleo_wb55rg, I face an issue on disconnect (while HRM notifications are on)
`[00:00:24.124,000] <wrn> bt_hci_core: Unhandled event 0x00 len 2: 0813  `
Then, impossible to reconnect.
Works fine on master

joerchan commented:
> Testing on nucleo_wb55rg, I face an issue on disconnect (while HRM notifications are on)
> `[00:00:24.124,000] <wrn> bt_hci_core: Unhandled event 0x00 len 2: 0813 `
> Then, impossible to reconnect.
> Works fine on master ([fae5d96](https://github.com/zephyrproject-rtos/zephyr/commit/fae5d96618632458e15c5c33dd81bb860126ad1b))
> 
> EDIT: I'm seeing the same behavior with disco_l475_iot1 (different BT controller, different HCI driver)

Thanks @erwango. Yes the HCI driver shouldn't matter. It is because the bt_recv_prio is doing net_buf_mem_pull on the buf, the state needs to be saved and restored.
It doesn't happen in the hci_driver.c because it doesn't re-use the buf.

Pushed a fix for this issue.

erwango commented:
> > Testing on nucleo_wb55rg, I face an issue on disconnect (while HRM notifications are on)
> > `[00:00:24.124,000] <wrn> bt_hci_core: Unhandled event 0x00 len 2: 0813 `
> > Then, impossible to reconnect.
> > Works fine on master ([fae5d96](https://github.com/zephyrproject-rtos/zephyr/commit/fae5d96618632458e15c5c33dd81bb860126ad1b))
> > EDIT: I'm seeing the same behavior with disco_l475_iot1 (different BT controller, different HCI driver)
> 
> Thanks @erwango. Yes the HCI driver shouldn't matter. It is because the bt_recv_prio is doing net_buf_mem_pull on the buf, the state needs to be saved and restored.
> It doesn't happen in the hci_driver.c because it doesn't re-use the buf.
> 
> Pushed a fix for this issue.

Ok, I'm not able to reproduce anymore.
I've tried to test more and it's working so far. 2 times, I had trouble to restart HRM notifications, but second request went through (not sure this is related of even a zephyr issue).

alexandru-porosanu-nxp commented:
> > @erwango @cvinayak @thoh-ot @alexandru-porosanu-nxp could you please test this with your respective Bluetooth platforms please? At the very least connection and disconnection multiple times, ideally with ongoing traffic.
> 
> Yep, I can do that; I'm thinking of using peripheral_hr (VEGA) & central_hr (NRF) and hack one of them to disconnect and reconnect at about 5-10s interval. If you have a better scenario in mind, let me know and I'll try to figure out how to do it...

I changed the scenario a bit and did about 10 connect/disconnect with NXP IoT toolbox to the VEGA board running peripheral_hr. So far, no issues.

jhedberg commented:
In general this does seem like the right way to solve the issue. However, could we consider some other name for the new bt_conn state? "Priority' is appropriate for a thread or context, but not a state IMO. Something like PRE_DISCONNECT, or DISCONNECT_1 and DISCONNECT_2 would make more sense to me.

jhedberg commented:
I think I'd also like to generalise this more. I didn't look through the driver changes, but you can't e.g. assume that the host won't change the parsing state of the net_buf, so net_buf_simple_save/restore may need to be used on the driver side.

That said, I don't think it's particularly clean to have the driver have to deal with this. The two-stage processing details would really be best to be left to the host. However, this will require even more redesign of the HCI driver interface. What we would need is a for the host to inject net_bufs to the "root" of the RX threaad. That way you could always feed the disconnect through bt_recv_prio() to the host, and the host would then re-inject it to itself through the RX thread. Another use case that might benefit from this is the ECDH emulation that hci_ecc.c does, i.e. it might make sense to have it feed the HCI events through the RX thread instead of its own.

joerchan commented:
> That said, I don't think it's particularly clean to have the driver have to deal with this. The two-stage processing details would really be best to be left to the host. However, this will require even more redesign of the HCI driver interface. What we would need is a for the host to inject net_bufs to the "root" of the RX threaad.

I was considering this, but this is complicated by the BT_RECV_IS_RX_THREAD option. All the drivers that don't select this one has the same handling for bt_recv and bt_recv_prio. We could move the processing of this to bt_recv in that case.
The drivers that select this option is the zephyr controller and the H4. Maybe we could remove this option from H4 to further simplify it.

joerchan commented on subsys/bluetooth/host/conn_internal.h:
```diff
@@ -212,6 +213,13 @@ void bt_conn_disconnect_all(u8_t id);
 /* Look up an existing connection */
 struct bt_conn *bt_conn_lookup_handle(u16_t handle);
 
+static inline bool bt_conn_is_handle_valid(struct bt_conn *conn)
```
The fact that the state is used to determine if the handle has been assigned should remain hidden IMO. I prefer to give conn as the argument.

jhedberg replied:
Agreed

joerchan committed:
Bluetooth: test: Print disconnect reason in hex value

Print the disconnect reason in hex value since the defines are in hex.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented:
@jhedberg @carlescufi Pushed a commit with suggestion for moving the bt_recv/bt_recv_prio handling to the Host when BT_RECV_IS_RX_THREAD is not defined.
If we agree to do this I can rebase it properly. Not exactly sure what should be done when this option is defined. The HCI driver is the one with the most complicated bt_recv/bt_recv_prio handling.

joerchan commented:
CC: @rugeGerritsen This affects the nRF BLE Controller.

joerchan commented:
> In general this does seem like the right way to solve the issue. However, could we consider some other name for the new bt_conn state? "Priority' is appropriate for a thread or context, but not a state IMO. Something like PRE_DISCONNECT, or DISCONNECT_1 and DISCONNECT_2 would make more sense to me.

I've been trying to come up with sensible names, it's not easy.
I would prefer to leave DISCONNECTED state name as is, since this is also the initial state.
The _1 and _2 versions I'm not sure if you meant for what is now DISCONNECTED_PRIO, and DISCONNECTED, or DISCONNECT, and DISCONNECTED_PRIO

Best I could come up with:
`BT_CONN_DISCONNECT_CMD` (HCI command has been sent)
`BT_CONN_DISCONNECT_EVT`  (HCI disconnect complete event has been received).
`BT_CONN_DISCONNECTED`
or possibly:
`BT_CONN_DISCONNECT_SENT`
`BT_CONN_DISCONNECT_RCVD`
`BT_CONN_DISCONNECTED`

@carlescufi @jhedberg Thoughts?

joerchan commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -202,7 +202,11 @@ static void bt_ipm_rx_thread(void)
+
+		if (hcievt->evtserial.type == HCI_ACL ||
+		    (hcievt->evtserial.type == HCI_EVT &&
+		     bt_hci_evt_is_normal(hcievt->evtserial.evt.evtcode))) {
```
Code moved to hci_core.c now.

joerchan commented on drivers/bluetooth/hci/rpmsg.c:
```diff
@@ -152,9 +152,13 @@ void bt_rpmsg_rx(u8_t *data, size_t len)
+		}
+
+		if (pkt_indicator == RPMSG_ACL ||
+		    (pkt_indicator == RPMSG_EVT && bt_hci_evt_is_normal(evt))) {
```
code moved to hci_core.c now.

joerchan commented on drivers/bluetooth/hci/spi.c:
```diff
@@ -388,7 +388,11 @@ static void bt_spi_rx_thread(void)
+			}
+
+			if (rxmsg[PACKET_TYPE] == HCI_ACL ||
+			    (rxmsg[PACKET_TYPE] == HCI_ACL &&
```
code moved to hci_core.c now, so the typo is resolved.

joerchan commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -111,7 +111,10 @@ static void rx_thread(void *p1, void *p2, void *p3)
 		if (frame[0] == H4_EVT && bt_hci_evt_is_prio(frame[1])) {
 			bt_recv_prio(buf);
-		} else {
+		}
```
Moved to hci_core.c now.

joerchan commented on include/drivers/bluetooth/hci_driver.h:
```diff
@@ -33,18 +33,17 @@ enum {
 
 /**
- * @brief Check if an HCI event is high priority or not.
+ * @brief Check if an HCI shall be given from high priority context.
```
Fixed

carlescufi commented on include/drivers/bluetooth/hci_driver.h:
```diff
@@ -33,18 +33,17 @@ enum {
- * to pass the event through the RX thread, however if it returns false
- * then this risks a deadlock.
+ * Helper for the HCI driver to know which events shall be given to
+ * bt_recv_prio() from another context than the RX thread (e.g. ISR).
```
```suggestion
 * bt_recv_prio(), which depending on the driver might be called from another context than the RX thread (e.g. ISR).
```

carlescufi commented on subsys/bluetooth/host/conn_internal.h:
```diff
@@ -9,6 +9,7 @@
  */
 typedef enum __packed {
```
Why is this packed by the way?

joerchan replied:
I think it is to make it an uint8_t instead of a uint32_t in the bt_conn definition, since it is declared as bt_conn_state_t state;
But not sure that is the original thought behind this.

carlescufi commented:
@jhedberg 

>I think I'd also like to generalise this more. I didn't look through the driver changes, but you can't e.g. assume that the host won't change the parsing state of the net_buf, so net_buf_simple_save/restore may need to be used on the driver side.

I think that in the case of an event that is **both** prio and normal we could guarantee that the host will leave it untouched.

>That said, I don't think it's particularly clean to have the driver have to deal with this. The two-stage processing details would really be best to be left to the host. 

In an ideal world, I completely agree. But the problem is the combined builds and the h4 driver, where the driver and stack are already tightly coupled by the fact that `CONFIG_BT_RECV_IS_RX_THREAD=y`

>That way you could always feed the disconnect through bt_recv_prio() to the host, and the host would then re-inject it to itself through the RX thread

That won't work in the combined build. You would be reordering data and events, because at the point you receive the disconnected event through `bt_recv_prio()` there might well be data and other events that are queued inside `hci_driver.c` in non-net_buf format yet, and that haven't been processed into the queue that ends up creating the sequence of calls to `bt_recv()`.

carlescufi reviewed (commented):
@joerchan 
My preference would be to unify the "disconnection in progress" state into a single one:

- `BT_CONN_DISCONNECTING`: Applies to both having called `bt_hci_disconnect` or having recieved the first disconnect event through `bt_recv_prio`
- `BT_CONN_DISCONNECTED`: Final disconnection complete processed

if that is not possible then I would suggest:

- `BT_CONN_DISCONNECTING_LOC` (or `BT_CONN_DISC_LOCAL`): Applies to having called `bt_hci_disconnect` 
- `BT_CONN_DISCONNECTING_REM` (or `BT_CONN_DISC_REMOTE`): Applies to having recieved the first disconnect event through `bt_recv_prio`
- `BT_CONN_DISCONNECTED`: Final disconnection complete processed

carlescufi commented:
I just realized, for the drivers where `CONFIG_BT_RECV_IS_RX_THREAD=n`, this breaks the priority handling of num complete (and now disconnect), which is still required. If those drivers just call `bt_recv()` and the RX thread is blocked then posting it to the rx queue won't help. You still need to be able to execute the processing of the event from another context, which in this case is whoever calls `bt_recv_prio()`

carlescufi commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -6355,49 +6411,39 @@ int bt_recv(struct net_buf *buf)
+		}
+
+		if (bt_hci_evt_is_normal(hdr->evt)) {
+			net_buf_put(&bt_dev.rx_queue, buf);
```
Is this safe after the `net_buf` has been processed by `hci_event_prio`? i.e. has its internal state not been modified?

joerchan replied:
It should be, because `hci_event_prio` will save the state and restore it before returning.

joerchan commented:
> @joerchan
> My preference would be to unify the "disconnection in progress" state into a single one:
> 
> * `BT_CONN_DISCONNECTING`: Applies to both having called `bt_hci_disconnect` or having recieved the first disconnect event through `bt_recv_prio`
> * `BT_CONN_DISCONNECTED`: Final disconnection complete processed
> 
> if that is not possible then I would suggest:
> 
> * `BT_CONN_DISCONNECTING_LOC` (or `BT_CONN_DISC_LOCAL`): Applies to having called `bt_hci_disconnect`
> * `BT_CONN_DISCONNECTING_REM` (or `BT_CONN_DISC_REMOTE`): Applies to having recieved the first disconnect event through `bt_recv_prio`
> * `BT_CONN_DISCONNECTED`: Final disconnection complete processed

The reason why I added a new state instead of re-using `BT_CONN_DISCONNECT` state is that that would introduce a no state change in `bt_conn_set_state` since the event is generated when initiating locally as well.
Using remote/local might be confusing since it might be related to who initiated the disconnect (hci status codes: `BT_HCI_ERR_REMOTE_USER_TERM_CONN` and  `BT_HCI_ERR_LOCALHOST_TERM_CONN`). So I think using cmd/evt or sent/rcvd would be better.

joerchan committed:
doc: relnotes: Add release notes for updated HCI driver event handling

Add release notes for updated HCI driver event handling for calling the
bt_recv() and bt_recv_prio() functions when delivering HCI events to
the host.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented:
@carlescufi @jhedberg Updated according to what we agreed on. The use of flags didn't turn out as clean as I would have wanted it, but maybe it is good enough.

jhedberg commented on include/drivers/bluetooth/hci_driver.h:
```diff
@@ -32,33 +32,39 @@ enum {
  *
- * @return true if the event can be processed in the RX thread, false
- *         if it cannot.
+ * @return true if the event shall be processed by bt_recv_prio, false
```
Needs updating to flags

joerchan replied:
Updated.

jhedberg reviewed (changes requested):
Looks generally good. Just one documentation issue spotted. I’ll still try to do another more thorough review later.

mtpr-ot commented:
The ADV resume commit f24c38c306 (in local 2.2.1 backport version) fixes our ADV race condition, +1 from me.

joerchan commented:
> The ADV resume commit [f24c38c](https://github.com/zephyrproject-rtos/zephyr/commit/f24c38c3069c1202641902fc9de8a71ce5752977) (in local 2.2.1 backport version) fixes our ADV race condition, +1 from me.

Great. I have created a separate bug-report in case others run into the issue, they can then see that it has been fixed.
